presto-verifier/src/main/java/com/facebook/presto/verifier/retry/RetryDriver.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.verifier.retry;
+
+import com.facebook.presto.verifier.framework.QueryException;
+import io.airlift.log.Logger;
+import io.airlift.units.Duration;
+
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Predicate;
+
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.lang.Math.pow;
+import static java.lang.Thread.currentThread;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class RetryDriver
+{
+    private static final Logger log = Logger.get(RetryDriver.class);
+
+    private final int maxAttempts;
+    private final Duration minBackoffDelay;
+    private final Duration maxBackoffDelay;
+    private final double scaleFactor;
+    private final Predicate<QueryException> retryPredicate;
+
+    public RetryDriver(
+            RetryConfig config,
+            Predicate<QueryException> retryPredicate)
+    {
+        this.maxAttempts = config.getMaxAttempts();
+        this.minBackoffDelay = requireNonNull(config.getMinBackoffDelay(), "minBackoffDelay is null");
+        this.maxBackoffDelay = requireNonNull(config.getMaxBackoffDelay(), "maxBackoffDelay is null");
+        this.scaleFactor = config.getScaleFactor();
+        this.retryPredicate = requireNonNull(retryPredicate, "retryPredicate is null");
+    }
+
+    public <V> V run(String callableName, RetryOperation<V> operation)
+    {
+        int attempt = 1;
+        while (true) {
+            try {
+                return operation.run();
+            }
+            catch (QueryException qe) {
+                if (attempt >= maxAttempts || !retryPredicate.test(qe)) {
+                    throw qe;
+                }
+
+                attempt++;
+                int delayMillis = (int) min(minBackoffDelay.toMillis() * pow(scaleFactor, attempt - 1), maxBackoffDelay.toMillis());
+                int jitterMillis = ThreadLocalRandom.current().nextInt(max(1, (int) (delayMillis * 0.1)));
+                log.debug(
```

Comment:
I believe we should not print the full stack trace - but print the fact that we are retrying a query.

Reply:
No stacktrace is printed for readability reason, only `callableName` and `qe.getErrorCode` is printed. Also, this is a debug-level logging.

Reply:
I wasn't clear here. 

I meant that we should have a `log.info` and actually put this in logs. And since we are not printing the stack trace - it should be fine. So a convoluted way to say - lets convert this to `log.info`

Reply:
Making it info-level logging will pollute the output. It's not very helpful for anyone who read the log to see a bunch of retry logging. Since we're going to log all failures in the output. We should build a dashboard so that release oncall read its signal easily.

Reply:
As I look into the existing logs from the current verifier (assuming that its logging is similar to the new one - if not please call that out), this is what we print in the logs:

```
2019-03-12T02:45:43.636-0700	INFO	main	stdout	Name: 20190305_220551_29858_fas8x

2019-03-12T02:45:43.636-0700	INFO	main	stdout	Schema (control): bizapps

2019-03-12T02:45:43.636-0700	INFO	main	stdout	Schema (test): bizapps

2019-03-12T02:45:43.636-0700	INFO	main	stdout	Valid: true

2019-03-12T02:45:43.636-0700	INFO	main	stdout	Query (test): SELECT "PC_hdloaner_onhand_02"."binsize" AS "binsize",

  "PC_hdloaner_onhand_02"."locator_code" AS "locator_code",

  "PC_hdloaner_onhand_02"."onhand_qty" AS "onhand_qty",

  "PC_hdloaner_onhand_02"."operating_system" AS "operating_system",

  "PC_hdloaner_onhand_02"."reorderpoint" AS "reorderpoint"

FROM (

  select * from hdloaner_onhand_02

) "PC_hdloaner_onhand_02"

2019-03-12T02:45:43.636-0700	INFO	main	stdout	Control Duration (secs): 301.826719127

2019-03-12T02:45:43.636-0700	INFO	main	stdout	   Test Duration (secs): 25.136472249999997

2019-03-12T02:45:43.636-0700	INFO	main	stdout	Control CPU (secs): 7469.8240000000005

2019-03-12T02:45:43.636-0700	INFO	main	stdout	   Test CPU (secs): 15275.392

2019-03-12T02:45:43.636-0700	INFO	main	stdout	         CPU Ratio: 2.0
```
per query.

So, if we retry twice and print 2 more lines, I am not sure how that pollutes the logs. Similar to how we are printing in the logs that we started the query, this is the result and the CPU comparison, we should also print failures.

Reply:
This is what I wanted to change - the new version will not output any of those logs. We run 10000 queries per suite, it would be impossible for human to process those log. Instead, we should only output information about the overall status and progress of the verifier. Each individual query event should be queried through xdb or presented through UI.